server/src/main/java/io/crate/planner/optimizer/rule/MoveLimitBeneathUnion.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.planner.optimizer.rule;
+
+import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
+import static io.crate.planner.optimizer.matcher.Patterns.source;
+
+import java.util.function.Function;
+
+import io.crate.expression.symbol.Literal;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.TransactionContext;
+import io.crate.planner.operators.Limit;
+import io.crate.planner.operators.LogicalPlan;
+import io.crate.planner.operators.Union;
+import io.crate.planner.optimizer.Rule;
+import io.crate.planner.optimizer.costs.PlanStats;
+import io.crate.planner.optimizer.matcher.Capture;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Pattern;
+import io.crate.types.DataTypes;
+
+public class MoveLimitBeneathUnion implements Rule<Limit> {
+
+    private final Capture<Union> unionCapture;
+    private final Pattern<Limit> pattern;
+
+    public MoveLimitBeneathUnion() {
+        this.unionCapture = new Capture<>();
+        this.pattern = typeOf(Limit.class)
+            .with(l -> l.isPushedDown() == false)
+            .with(source(), typeOf(Union.class).capturedAs(unionCapture));
+    }
+
+    @Override
+    public Pattern<Limit> pattern() {
+        return pattern;
+    }
+
+    @Override
+    public LogicalPlan apply(Limit limit,
+                             Captures captures,
+                             PlanStats planStats,
+                             TransactionContext txnCtx,
+                             NodeContext nodeCtx,
+                             Function<LogicalPlan, LogicalPlan> resolvePlan) {
+        Union union = captures.get(unionCapture);
+        // We cannot push the offset down to the sources, otherwise
+        // we may loose rows, therefore we create a new limit where
+        // the offset is added to the limit.
```

Comment:
I think that this is not correct, what happens if we have an ORDER BY on the union, or if we have subselects with order by:
```
SELECT * FROM (SELECT * FROM t1 ORDER BY a) t1 UNION ALL (SELECT * FROM t2 ORDER BY b) t2
```
I think there should be a global Limit that that applies offset + limit .

Reply:
The offset is removed and the limit is extended by the offset. So in case of a order by on the union, the order by is not affected.

Reply:
> I think there should be a global Limit that that applies offset + limit .

This is exactly what is happening here.

Reply:
```
var limitVal = DataTypes.LONG.sanitizeValue(l.value());
var offSetVal = DataTypes.LONG.sanitizeValue(o.value());
var newLimit = offSetVal + limitVal;
```

Reply:
Thx, I was confused by the comment, apologies.